internal/sync/syncer.go

```diff
@@ -0,0 +1,311 @@
+func KeyOfGitSync(gitSync *v1alpha1.GitSync) string {
+	return fmt.Sprintf("%s/%s", gitSync.Namespace, gitSync.Name)
+}
+
+// NewSyncer returns a Synchronizer instance.
+func NewSyncer(client client.Client, config *rest.Config, rawConfig *rest.Config, kubectl kubeutil.Kubectl, opts ...Option) *Syncer {
+	watcherOpts := defaultOptions()
+	for _, opt := range opts {
+		if opt != nil {
+			opt(watcherOpts)
+		}
+	}
+	stateCache := NewLiveStateCache(config)
+	w := &Syncer{
+		client:      client,
+		config:      config,
+		rawConfig:   rawConfig,
+		kubectl:     kubectl,
+		options:     watcherOpts,
+		gitSyncMap:  make(map[string]*list.Element),
+		gitSyncList: list.New(),
+		lock:        new(sync.RWMutex),
+		stateCache:  stateCache,
+	}
+	return w
+}
+
+// Contains returns if the synchronizer contains the key (namespace/name).
+func (s *Syncer) Contains(key string) bool {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	_, ok := s.gitSyncMap[key]
+	return ok
+}
+
+// Length returns how many GitSync objects are watched
+func (s *Syncer) Length() int {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	return s.gitSyncList.Len()
+}
+
+// StartWatching put a key (namespace/name) into the synchronizer
+func (s *Syncer) StartWatching(key string) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if _, ok := s.gitSyncMap[key]; !ok {
+		s.gitSyncMap[key] = s.gitSyncList.PushBack(key)
+	}
+}
+
+// StopWatching stops watching on the key (namespace/name)
+func (s *Syncer) StopWatching(key string) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if e, ok := s.gitSyncMap[key]; ok {
+		_ = s.gitSyncList.Remove(e)
+		delete(s.gitSyncMap, key)
+	}
+}
+
+// Start function starts the synchronizer worker group.
+// Each worker keeps picking up tasks (which contains GitSync keys) to sync the resources.
+func (s *Syncer) Start(ctx context.Context) error {
+	log := logging.FromContext(ctx).Named("synchronizer")
+	log.Info("Starting synchronizer...")
+	keyCh := make(chan string)
+	ctx, cancel := context.WithCancel(logging.WithLogger(ctx, log))
+	defer cancel()
+
+	s.stateCache.Init()
+
+	// Worker group
+	for i := 1; i <= s.options.workers; i++ {
+		go s.run(ctx, i, keyCh)
+	}
+
+	// Function assign() moves an element in the list from the front to the back,
+	// and send to the channel so that it can be picked up by a worker.
+	assign := func() {
+		s.lock.Lock()
+		defer s.lock.Unlock()
+		if s.gitSyncList.Len() == 0 {
+			return
+		}
+		e := s.gitSyncList.Front()
+		if key, ok := e.Value.(string); ok {
+			s.gitSyncList.MoveToBack(e)
+			keyCh <- key
+		}
+	}
+
+	// Following for loop keeps calling assign() function to assign watching tasks to the workers.
+	// It makes sure each element in the list will be assigned every N milliseconds.
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("Shutting down synchronizer job assigner.")
+			return nil
+		default:
+			assign()
+		}
+		// Make sure each of the key will be assigned at most every N milliseconds.
```

Comment:
This is neat logic. I see here you wrote "the key will be assigned at most every N milliseconds" - that makes sense since the Workers could get busy and the channel could get blocked. I wonder if the comment above should also say "at most"?

Reply:
Now that I say that...say the channel has been blocked for awhile...then we will sleep here even though the task may be due, right? (thinking if we'd ever want to do anything about that in the future or not. I suppose this is an undesirable state of too much work for the Controller with not enough workers, which means somebody should probably tune their Controller for more workers and more CPU. And maybe our sleeping here isn't such a bad thing)

Reply:
yeah I think that is an indication of more workers are needed. but lets keep an eye on this to see if improvements needed especially with auto healing.